zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
-// the supplied bounding rectangle
-double avgOfDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect)
+// Find the median of a continuous cv::Mat
+float findMedianOfMat(cv::Mat mat) {
+	float median = 0;
+	if (mat.isContinuous()) {
+		// copy matrix data to a vector
+		std::vector<float> vec;
+		mat = mat.reshape(0, 1);
+		mat.copyTo(vec);
+		// remove 255 (when this is called, values that are masked are 255)
+		vec.erase(std::remove(vec.begin(), vec.end(), 255), vec.end());
+		// sort vector
+		std::sort(vec.begin(), vec.end()); // forgot this earlier
+		if ((vec.size() % 2) != 0) { // if odd
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median = vec[vec.size()/2];
+		} else { // if even
+			auto it = vec.begin() + vec.size()/2;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			it = vec.begin() + vec.size()/2 - 1;
+			std::nth_element(vec.begin(), it, vec.end());
+			median += vec[vec.size()/2];
+			median = ((float)median/2.0d);
+		}
+	}
+	return median;
+}
+
+// K_MEANS = k-means, Forgy method of initializing (k-means++ would likely be better, but Forgy is working ok for now)
+// CONTOURS = find depth by finding contours and taking the lowest value of the biggest contour
+// CONTOURS_NON_ADAPTIVE = contours, but using thresholding based on the median instead of adaptive thresholding
+enum DepthCalculationAlgorithm { K_MEANS, CONTOURS, CONTOURS_NON_ADAPTIVE };
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using contour finding
+double contoursDepthMat(const cv::Mat& depth_, const cv::Rect& bound_rect, bool debug = false, bool adaptive = true) {
+	if (bound_rect.size().area() == 0) { // if the ROI is zero, return -1 (no depth)
+		return -1;
+	}
+
+	// Crop depth to region of interest
+	cv::Mat depth = depth_(bound_rect);
+
+	// set very large outliers and nan to 0 so they can be removed later
+	float nan_ = std::numeric_limits<float>::quiet_NaN();
+	cv::Mat inf = depth>=900;
+	cv::Mat neg_inf = depth<=-900;
+	cv::Mat nan = depth==nan_;
+	cv::Mat neg_nan = depth==-nan_;
+	depth.setTo(0, inf);
+	depth.setTo(0, neg_inf);
+	depth.setTo(0, nan);
+	depth.setTo(0, neg_nan);
+
+	if (debug) {
+		double min, max;
+		cv::minMaxLoc(depth, &min, &max);
+		ROS_INFO_STREAM("min: " << min << ", max: " << max);
+		cv::Mat dest;
+		depth.copyTo(dest);
+		cv::normalize(dest, dest, 0, 255, cv::NORM_MINMAX);
+		dest.convertTo(dest, CV_8UC1);
+		cv::imshow("Depth", dest);
+		cv::waitKey(1);
+	}
+
+	// convert depth to a 0-255 grayscale image (for contour finding)
+	cv::Mat depthDifferentFormat;
+	cv::Mat zeros = depth==0; // no depth is 0 with the ZED. May also need to check for inf and nan.
+	depth.copyTo(depthDifferentFormat);
+	cv::normalize(depthDifferentFormat, depthDifferentFormat, 0, 128, cv::NORM_MINMAX); // 0-128 because outliers will be 255 (and we want a clear background)
```

Comment:
Does this work if the copy is removed and depth is passed as the 1st argument to normalize?

Reply:
I don't think so -- it uses the normalized version to detect contours, and then at the end of the code it uses the original depth data to find the actual depth.

Reply:
Ah, missed that. My bad, appreciate checking it to be sure.